Improved method for chessboard detection

mshabunin opened the pull request:
Ported from #4046 and #6851 by @HHI-VIT-IMC

>Modified and improved the method for chessboard detection. It is now faster and detects chessboards under difficult lighting condition as well as when the chessboard has strong out of plane rotations

Original PR has been transferred from 2.4 branch to master.

HHI-VIT-IMC commented:
Am I right that now the improved chessboard detection is part of the next OpenCV release? If so, many thanks for the support.

mshabunin commented:
@HHI-VIT-IMC , it will become part of new release (3.2) after this PR is merged.

Please, note that I've also reverted `min_aspect_ratio` and `max_aspect_ratio` in `icvGetQuadrangleHypotheses`, becasue otherwise the chessboard is detected on images without it.
I've also fixed the test which checked this behaviour (ported changes from #4187 and opencv/opencv_extra#261).

mshabunin commented:
@vpisarev , could you please take a look?

vpisarev commented:
@mshabunin, thanks for preparing this PR! Let's put it in so that users can test it thoroughly. Also, I would not mind if we put some extra test images with and without chessboards. :+1:

alalek commented on modules/calib3d/src/calibinit.cpp:
```diff
@@ -191,38 +195,204 @@ static void icvRemoveQuadFromGroup(CvCBQuad **quads, int count, CvCBQuad *q0);
-            i++;
-        }
+      iIdx = i+ii;
+      if (iIdx > 0 && iIdx < 256)
```
```.diff
-iIdx > 0
+iIdx >= 0
```

#11591

alalek commented on modules/calib3d/src/calibinit.cpp:
```diff
@@ -191,38 +195,204 @@ static void icvRemoveQuadFromGroup(CvCBQuad **quads, int count, CvCBQuad *q0);
+  std::vector<int> piHistIntensity(iNumBins, 0);
+  std::vector<int> piHistSmooth(iNumBins, 0);
+  std::vector<int> piHistGrad(iNumBins, 0);
+  std::vector<int> piAccumSum(iNumBins, 0);
```
`piAccumSum` is calculated, but not used anywhere.

alalek commented on modules/calib3d/src/calibinit.cpp:
```diff
@@ -191,38 +195,204 @@ static void icvRemoveQuadFromGroup(CvCBQuad **quads, int count, CvCBQuad *q0);
+  {
+    iThresh = (piMaxPos[0] + piMaxPos[1])/2;
+  }
+  else // iCntMaxima >= 3
```
Case `iCntMaxima == 0` is lost and incorrectly handled by "iCntMaxima >= 3" code.

#11590

alalek commented on modules/calib3d/src/calibinit.cpp:
```diff
@@ -191,38 +195,204 @@ static void icvRemoveQuadFromGroup(CvCBQuad **quads, int count, CvCBQuad *q0);
+    if ( piMaxPos[iIdxBGMax] >= 250 && iIdxBGMax < iCntMaxima )
+    {
+      iIdxBGMax++;
+      iMaxVal = piHistIntensity[piMaxPos[iIdxBGMax]];
```
Pre-condition above should avoid `piMaxPos` buffer overrun on this line:
```.diff
-iIdxBGMax < iCntMaxima
+iIdxBGMax + 1 < iCntMaxima
```

#11590

alalek commented on modules/calib3d/src/calibinit.cpp:
```diff
@@ -300,213 +464,258 @@ int cvFindChessboardCorners( const void* arr, CvSize pattern_size,
-                    int thresh_level = cvRound( mean - 10 );
-                    thresh_level = MAX( thresh_level, 10 );
+      //USE BINARY IMAGE COMPUTED USING icvBinarizationHistogramBased METHOD
+      cvDilate( thresh_img_new, thresh_img_new, 0, 1 );
```
cvFindContours() (from icvGenerateQuads) modifies image, which becomes changed for next dilation iterations.

#11589
